src/main/java/bio/terra/common/stairway/StairwayComponent.java

```diff
@@ -0,0 +1,138 @@
+            .workQueueProjectId(getDefaultProjectId())
+            .enableWorkQueue(kubernetesComponent.isInKubernetes())
+            .stairwayHook(new TracingHook());
+    try {
+      stairway = builder.build();
+    } catch (StairwayExecutionException e) {
+      throw new IllegalArgumentException("Failed to build Stairway.", e);
+    }
+  }
+
+  public void initialize() {
+    logger.info("Initializing Stairway...");
+    logger.info("stairway username {}", stairwayJdbcConfiguration.getUsername());
+    try {
+      // TODO(PF-161): Determine if Stairway and buffer database migrations need to be coordinated.
+      List<String> recordedStairways =
+          stairway.initialize(
+              stairwayJdbcConfiguration.getDataSource(),
+              stairwayConfiguration.isForceCleanStart(),
+              stairwayConfiguration.isMigrateUpgrade());
+
+      kubeService.startPodListener(stairway);
+
+      // Lookup all of the stairway instances we know about
+      Set<String> existingStairways = kubeService.getPodSet();
+      List<String> obsoleteStairways = new LinkedList<>();
+
+      // Any instances that stairway knows about, but we cannot see are obsolete.
+      for (String recordedStairway : recordedStairways) {
+        if (!existingStairways.contains(recordedStairway)) {
+          obsoleteStairways.add(recordedStairway);
+        }
+      }
+
+      // Add our own pod name to the list of obsolete stairways. Sometimes Kubernetes will
+      // restart the container without redeploying the pod. In that case we must ask
+      // Stairway to recover the flights we were working on before being restarted.
+      obsoleteStairways.add(kubeService.getPodName());
+
+      logger.info(
+          "existingStairways: {}. obsoleteStairways: {}", existingStairways, obsoleteStairways);
+      // Recover and start stairway - step 3 of the stairway startup sequence
+      stairway.recoverAndStart(obsoleteStairways);
+    } catch (StairwayException | InterruptedException e) {
+      status.compareAndSet(Status.INITIALIZING, Status.ERROR);
+      throw new RuntimeException("Error starting Stairway", e);
+    }
+    status.compareAndSet(Status.INITIALIZING, Status.OK);
+  }
+
+  /** Stop accepting jobs and shutdown stairway. Returns true if successful. */
+  public boolean shutdown() throws InterruptedException {
+    status.set(Status.SHUTDOWN);
+    logger.info("Request Stairway shutdown");
+    boolean shutdownSuccess =
+        stairway.quietDown(
+            stairwayConfiguration.getQuietDownTimeout().toMillis(), TimeUnit.MILLISECONDS);
+    if (!shutdownSuccess) {
+      logger.info("Request Stairway terminate");
+      shutdownSuccess =
+          stairway.terminate(
+              stairwayConfiguration.getTerminateTimeout().toMillis(), TimeUnit.MILLISECONDS);
+    }
+    logger.info("Finished Stairway shutdown?: {}", shutdownSuccess);
+    return shutdownSuccess;
+  }
+
+  public Stairway get() {
```

Comment:
@gjuggler so this becomes a `Provider<Stairway>` that's separately provided by the new config class and autowired?

Reply:
We actually can't make `Stairway` into a bean if we want to configure hooks and the context in the initialize method. This also means that the KubePodListener can't be provided as a bean.

Reply:
Why does Stairway need to be a bean? Injection is a convenience, not a requirement.

I don't think KubePodListener should be bean. It is only used by KubeService, so never needs to be injected.

Reply:
Good point. It could be an inner class, but it feels like too much code for that.